### PR TITLE
Reimplement java backend builtin list

### PR DESCRIPTION
**Pull request by yilongli**
Provides important support for proving several kernelc examples that mention I/O cells.

### CODE REVIEW

**Comment by yilongli** [comment]
@andreistefanescu @dwightguth please review

**Comment by yilongli** [comment]
Jenkins: retest this please

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java [hunk 1]
```diff
@@ -2,81 +2,92 @@
...
-    public BuiltinList(Variable frame, int removeLeft, int removeRight, java.util.Collection<Term> elementsLeft, java.util.Collection<Term> elementsRight) {
-        super(frame, Kind.KITEM);
-        this.elementsLeft = new ImprovedArrayDeque<Term>(elementsLeft);
-        if (frame == null) {
-            assert removeLeft == 0 && removeRight == 0 : "cannot remove from an empty base";
-            this.elementsLeft.addAll(elementsRight);
-            this.elementsRight = new ImprovedArrayDeque<Term>();
+    private final ImmutableList<Term> elementsLeft;
+    private final ImmutableList<Term> elementsRight;
+    private final ImmutableList<Term> baseTerms;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:36]: is this supposed to be like, something where baseTerms and baseTermTypes match up one to one? Maybe you should make a BaseTerm class that contains both of those fields, if you're going to do it this way?

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:36]: Yes. I have thought of this. However, the `baseTermTypes` is not used anywhere right now. And I am not very sure about how it's going to be used in the unification algorithm. So I would like to postpone this decision until I start to prove the examples involving IO cells. I have added a TODO item to keep track of this.

**Comment by yilongli** [comment]
This is the second time I get this error in this pull request. It's really weird. http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/482/testReport/junit/_samples_kernelc_kernelc/k/kernelc_k/

**Comment by yilongli** [comment]
Jenkins: retest this please

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java [hunk 2]
```diff
@@ -62,21 +62,25 @@ public Variable frame() {
...
-         * frame stuff becomes out-dated now
-         */
-        return size() == 0 && !hasFrame();
+        return size() == 0 && isConcreteCollection();
     }
 
     /**
      * Returns the size of the contents of this {@code Collection}.
      *
+     * TODO(YilongL): rename to concreteSize?
```
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java]: yes, please

**Comment by andreistefanescu** [comment]
Looks good overall.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java (hunk 1 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:36]: Today I realize that if I combine these two lists together,  the signature of `BuiltinList#baseTerms()` will be inconsistent with other builtin collections. Besides, it requires to change many places in the code base. Therefore, although I have created a new class `BaseTerm`, I decide not to combine the two lists. Instead, I add a new method `BaseTerm getBaseTerm(int index)`, which is only used in the `SymbolicUnifier`.

**Comment by yilongli** [comment]
@andreistefanescu I have implemented a naive and very limited algorithm for unifying two BuiltinLists. I try to modify the code in `SymbolicConstraint` by following what you did for BuiltinMaps. However, I do not fully understand it. Please review again.

**Comment by yilongli** [comment]
`[Error] Internal: Stratego rewriting failed` again...
http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/485/testReport/junit/_samples_java_rewrite_engine_tutorial_2_languages_1_simple_1_untyped_simple-untyped/k/simple_untyped_k/

**Comment by yilongli** [comment]
Jenkins: retest this please

**Comment by yilongli** [comment]
I have been able to use the new `BuiltinList` to prove 4 more tree examples. Please review. I will try to prove the last tree example (`tree_to_list_iterative.c`) tomorrow.

**Comment by andreistefanescu** [comment]
Great job Yilong!

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java (hunk 1 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:36]: okay

**Pull request merged**